text/0078-merkleized-metadata.md

```diff
@@ -0,0 +1,393 @@
+
+The `SignedExtensionMetadata` provides information about a signed extension:
+
+- `identifier`: The `identifier` of the signed extension. An `identifier` is required to be unique in the Polkadot ecosystem as otherwise extrinsics are maybe built incorrectly.
+- `included_in_extrinsic`: The type that will be included in the extrinsic by this signed extension.
+- `included_in_signed_data`: The type that will be included in the signed data by this signed extension.
+
+### Type Information
+
+As SCALE is not self descriptive like JSON, a decoder always needs to know the format of the type to decode it properly. This is where the type information comes into play. The format of the extrinsic is fixed as described above and `ExtrinsicMetadata`  provides information on which type information is required for which part of the extrinsic. So, offline wallets only need access to the actual type information. It is a requirement that the type information can be chunked into logical pieces to reduce the amount of data that is sent to the offline wallets for decoding the extrinsics. So, the type information is structured in the following way:
+
+```rust
+struct Type {
+    path: Vec<String>,
+    type_def: TypeDef,
+    type_id: Compact<u32>,
+}
+
+enum TypeDef {
+    Composite(Vec<Field>),
+    Enumeration(EnumerationVariant),
+    Sequence(TypeRef),
+    Array(Array),
+    Tuple(Vec<TypeRef>),
+    BitSequence(BitSequence),
+}
+
+struct Field {
+    name: Option<String>,
+    ty: TypeRef,
+    type_name: Option<String>,
+}
+
+struct Array {
+    len: u32,
+    type_param: TypeRef,
+}
+
+struct BitSequence {
+    num_bytes: u8,
+    least_significant_bit_first: bool,
+}
+
+struct EnumerationVariant {
+    name: String,
+    fields: Vec<Field>,
+    index: Compact<u32>,
+}
+
+enum TypeRef {
+    Bool,
+    Char,
+    Str,
+    U8,
+    U16,
+    U32,
+    U64,
+    U128,
+    U256,
+    I8,
+    I16,
+    I32,
+    I64,
+    I128,
+    I256,
+    CompactU8,
+    CompactU16,
+    CompactU32,
+    CompactU64,
+    CompactU128,
```

Comment:
Shall we add `CompactU256` to make it 100% future-proof? It might happen at some point.